src/main/scala/com/avast/scala/hashes/circe/package.scala

```diff
@@ -1,12 +1,31 @@
 package com.avast.scala.hashes
 
-import io.circe.{Decoder, Encoder}
+import io.circe._
+
+import scala.reflect.ClassTag
+import scala.util.control.NonFatal
 
 package object circe {
-  implicit val Sha256Decoder = Decoder.decodeString.map(Sha256(_))
-  implicit val Sha256Encoder = Encoder.encodeString.contramap((s: Sha256) => s.toString)
-  implicit val MD5Decoder = Decoder.decodeString.map(MD5(_))
-  implicit val MD5Encoder = Encoder.encodeString.contramap((h: MD5) => h.toString)
-  implicit val Sha1Decoder = Decoder.decodeString.map(Sha1(_))
-  implicit val Sha1Encoder = Encoder.encodeString.contramap((s: Sha1) => s.toString)
+  implicit val Sha256Decoder: Decoder[Sha256] = prepareDecoder(Sha256(_))
+  implicit val Sha256Encoder: Encoder[Sha256] = Encoder.encodeString.contramap((s: Sha256) => s.toString)
+  implicit val MD5Decoder: Decoder[MD5] = prepareDecoder(MD5(_))
+  implicit val MD5Encoder: Encoder[MD5] = Encoder.encodeString.contramap((h: MD5) => h.toString)
+  implicit val Sha1Decoder: Decoder[Sha1] = prepareDecoder(Sha1(_))
+  implicit val Sha1Encoder: Encoder[Sha1] = Encoder.encodeString.contramap((s: Sha1) => s.toString)
+
+  private def prepareDecoder[A: ClassTag](parse: String => A): Decoder[A] =
+    (c: HCursor) => {
+      c.value.as[String].flatMap { s =>
```

Comment:
IMHO the `.as[String]` can cause an exception. So we could use the same as `decodeString` uses:
```scala
implicit final val decodeString: Decoder[String] = new Decoder[String] {
    final def apply(c: HCursor): Result[String] = c.value match {
      case Json.JString(string) => Right(string)
      case _                    => Left(DecodingFailure("String", c.history))
    }
  }
```
Or maybe just use `toString()` instead of `as[String]` would work well.

Reply:
`.as[String]` is Circe json String decoder, so it's correct usage.

Reply:
... and `Json.JString` is private class :-(